go/cli/mcap/cmd/filter.go

```diff
@@ -0,0 +1,317 @@
+	}
+
+	includeTopics, err := compileMatchers(filterIncludeTopics)
+	if err != nil {
+		return nil, err
+	}
+	opts.includeTopics = includeTopics
+
+	excludeTopics, err := compileMatchers(filterExcludeTopics)
+	if err != nil {
+		return nil, err
+	}
+	opts.excludeTopics = excludeTopics
+	return opts, nil
+}
+
+// filterCmd represents the filter command
+var filterCmd = &cobra.Command{
+	Use:   "filter",
+	Short: "copy some filtered MCAP data to a new file",
+	Long: `This subcommand filters an MCAP by topic and time range to a new file.
+When multiple regexes are used, topics that match any regex are included (or excluded).
+
+usage example:
+  mcap filter in.mcap -o out.mcap -y /diagnostics -y /tf -y /camera_(front|back)`,
+	Run: func(cmd *cobra.Command, args []string) {
+		filterOptions, err := buildFilterOptions()
+		if err != nil {
+			die("configuration error: %s", err)
+		}
+		var reader io.Reader
+		if len(args) == 0 {
+			stat, err := os.Stdin.Stat()
+			if err != nil {
+				die("failed to check stdin state: %s", err)
+			}
+			if stat.Mode()&os.ModeCharDevice == 0 {
+				reader = os.Stdin
+			} else {
+				die("supply a file to read")
+			}
+		} else {
+			close, newReader, err := utils.GetReader(context.Background(), args[0])
+			if err != nil {
+				die("failed to open source for reading: %s", err)
+			}
+			defer func() {
+				if closeErr := close(); closeErr != nil {
+					die("error closing read source: %s", closeErr)
+				}
+			}()
+			reader = newReader
+		}
+
+		var writer io.Writer
+		if filterOutput == "" {
+			if !utils.StdoutRedirected() {
+				die("Binary output can screw up your terminal. Supply -o or redirect to a file or pipe")
+			}
+			writer = os.Stdin
+		} else {
+			newWriter, err := os.Create(filterOutput)
+			if err != nil {
+				die("failed to open %s for writing: %s", filterOutput, err)
+			}
+			writer = newWriter
+		}
+
+		err = filter(reader, writer, filterOptions)
+		if err != nil {
+			die(error.Error(err))
+		}
+	},
+}
+
+func compileMatchers(regexStrings []string) ([]regexp.Regexp, error) {
+	var matchers []regexp.Regexp
+
+	for _, regexString := range regexStrings {
+		// auto-surround with ^$ if not specified.
+		if regexString[:1] != "^" {
+			regexString = "^" + regexString
+		}
+		if regexString[len(regexString)-1:] != "$" {
+			regexString = regexString + "$"
+		}
+		regex, err := regexp.Compile(regexString)
+		if err != nil {
+			return nil, err
+		}
+		matchers = append(matchers, *regex)
+	}
+	return matchers, nil
+}
+
+type markableSchema struct {
+	*mcap.Schema
+	written bool
+}
+type markableChannel struct {
+	*mcap.Channel
+	written bool
+}
+
+func filter(
+	r io.Reader,
+	w io.Writer,
+	opts *filterOpts,
+) error {
+	lexer, err := mcap.NewLexer(r)
+	if err != nil {
+		return err
+	}
+	mcapWriter, err := mcap.NewWriter(w, &mcap.WriterOptions{
+		Compression: opts.compressionFormat,
+	})
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = mcapWriter.Close()
```

Comment:
I think this probably should either log the error here in the defer, or do a little more work to overwrite the returned error. Currently the error is just getting dropped.

To properly return the close error from the function you would want to change the function to return a named variable `(err error)` and then do some similar business to https://github.com/foxglove/data-platform/blob/main/stream-server/merging/mcap0_to_bag.go#L35,L45.